api/src/services/assets.ts

```diff
@@ -50,6 +52,83 @@ export class AssetsService {
 		this.filesService = new FilesService({ ...options, accountability: null });
 	}
 
+	async getFileHierarchy(rootFolder: string) {
+		const fileTree = this.knex
+			.withRecursive('folders', ['id', 'name', 'parent'], (qb) => {
+				qb.select('id', 'name', 'parent')
+					.from('directus_folders')
+					.where('id', rootFolder)
+					.unionAll((qb) => {
+						qb.select(
+							'directus_folders.id',
+							this.knex.raw(`CONCAT(folders.name, '/', directus_folders.name)`),
+							'directus_folders.parent',
+						)
+							.from('directus_folders')
+							.join('folders', 'folders.id', 'directus_folders.parent');
+					});
+			})
+			.select<{ id: string; folder: string }[]>('directus_files.id', 'folders.name as folder')
+			.from('directus_files')
+			.rightJoin('folders', 'folders.id', 'directus_files.folder');
+
+		const results = await fileTree;
+
+		return results;
+	}
+
+	async getZip(files: Files, skipUnaccessable = false) {
+		const archive = archiver('zip');
+
+		const complete = async () => {
+			const fileDeduper = new FileDeduper();
+			const storage = await getStorage();
+
+			for (const { id, folder } of files) {
+				if (!isValidUuid(id)) throw new ForbiddenError();
+				let accessError;
+
+				if (this.accountability) {
+					try {
+						await validateAccess(
+							{
+								accountability: this.accountability,
+								action: 'read',
+								collection: 'directus_files',
+								primaryKeys: [id],
+							},
+							{ knex: this.knex, schema: this.schema },
+						);
+					} catch (error) {
+						accessError = error;
+					}
+
+					if (!skipUnaccessable && accessError) {
+						throw accessError;
+					}
+				}
+
+				const file = (await this.filesService.readOne(id)) as File;
```

Comment:
This file service is still using admin and will provide more permissions than the user has. The above validateAccess check only ensures they have access to the item, not what fields they have permission to.

Reply:
I believe `storage` and `filename_disk` should be pulled in irrespective of permissions, similar to how we would do it for getAsset once fixed. As we do not expose it back to the user and is required for the operation.